zebra-network/src/peer/connection.rs

```diff
@@ -369,454 +516,226 @@ where
-                            handler: Handler::FindBlocks,
-                            tx,
-                            span,
-                        },
-                        None,
-                    )),
-                    Err(e) => Err((e, tx)),
+                    Ok(()) => Transition::AwaitResponse {
+                        handler: Handler::FindBlocks,
+                        tx,
+                        span,
+                    },
+                    Err(e) => Transition::ExitResponse { e: e.into(), tx },
                 }
             }
-            (AwaitingRequest, FindHeaders { known_blocks, stop }) => {
+            FindHeaders { known_blocks, stop } => {
                 match self
                     .peer_tx
                     .send(Message::GetHeaders { known_blocks, stop })
                     .await
                 {
-                    Ok(()) => Ok((
-                        AwaitingResponse {
-                            handler: Handler::FindHeaders,
-                            tx,
-                            span,
-                        },
-                        None,
-                    )),
-                    Err(e) => Err((e, tx)),
-                }
-            }
-            (AwaitingRequest, MempoolTransactions) => {
-                match self.peer_tx.send(Message::Mempool).await {
-                    Ok(()) => Ok((
-                        AwaitingResponse {
-                            handler: Handler::MempoolTransactions,
-                            tx,
-                            span,
-                        },
-                        None,
-                    )),
-                    Err(e) => Err((e, tx)),
+                    Ok(()) => Transition::AwaitResponse {
+                        handler: Handler::FindHeaders,
+                        tx,
+                        span,
+                    },
+                    Err(e) => Transition::ExitResponse { e: e.into(), tx },
                 }
             }
-            (AwaitingRequest, PushTransaction(transaction)) => {
+            MempoolTransactions => match self.peer_tx.send(Message::Mempool).await {
+                Ok(()) => Transition::AwaitResponse {
+                    handler: Handler::MempoolTransactions,
+                    tx,
+                    span,
+                },
+                Err(e) => Transition::ExitResponse { e: e.into(), tx },
+            },
+            PushTransaction(transaction) => {
                 match self.peer_tx.send(Message::Tx(transaction)).await {
-                    Ok(()) => Ok((AwaitingRequest, Some(tx))),
-                    Err(e) => Err((e, tx)),
+                    Ok(()) => {
+                        // Since we're not waiting for further messages, we need to
+                        // send a response before dropping tx.
+                        let _ = tx.send(Ok(Response::Nil));
+                        Transition::AwaitRequest
```

Comment:
Can we move this common code into the transition function, so every message that doesn't expect a response gets the same treatment?